pkg/storegateway/indexheader/reader_pool.go

```diff
@@ -88,26 +92,46 @@ func (l lazyLoadedHeadersSnapshot) persist(persistDir string) error {
 
 // NewReaderPool makes a new ReaderPool. If lazy-loading is enabled, NewReaderPool also starts a background task for unloading idle Readers and persisting a list of loaded Readers to disk.
 func NewReaderPool(logger log.Logger, lazyReaderEnabled bool, lazyReaderIdleTimeout time.Duration, metrics *ReaderPoolMetrics, lazyLoadedSnapshotConfig LazyLoadedHeadersSnapshotConfig) *ReaderPool {
-	p := newReaderPool(logger, lazyReaderEnabled, lazyReaderIdleTimeout, metrics)
+	var snapshot *lazyLoadedHeadersSnapshot
+	if lazyReaderEnabled && lazyLoadedSnapshotConfig.IndexHeadersEagerLoadingStartupEnabled {
+		lazyLoadedSnapshotFileName := filepath.Join(lazyLoadedSnapshotConfig.Path, lazyLoadedHeadersListFile)
+		var err error
+		snapshot, err = loadLazyLoadedHeadersSnapshot(lazyLoadedSnapshotFileName)
+		if err != nil {
+			level.Warn(logger).Log("msg", "loading lazy-loaded index header failed; removing the file", "file", lazyLoadedSnapshotFileName, "err", err)
+			err := os.Remove(lazyLoadedSnapshotFileName)
+			if err != nil {
+				level.Warn(logger).Log("msg", "removing lazy-loaded index header failed", "file", lazyLoadedSnapshotFileName, "err", err)
+			}
+		}
+	}
+
+	p := newReaderPool(logger, lazyReaderEnabled, lazyReaderIdleTimeout, lazyLoadedSnapshotConfig.IndexHeadersEagerLoadingStartupEnabled, metrics, snapshot)
 
 	// Start a goroutine to close idle readers (only if required).
 	if p.lazyReaderEnabled && p.lazyReaderIdleTimeout > 0 {
 		checkFreq := p.lazyReaderIdleTimeout / 10
 
 		go func() {
-			tickerLazyLoadPersist := time.NewTicker(time.Minute)
-			defer tickerLazyLoadPersist.Stop()
-
 			tickerIdleReader := time.NewTicker(checkFreq)
 			defer tickerIdleReader.Stop()
 
+			var lazyLoadC <-chan time.Time
+
+			if p.eagerLoadReaderEnabled {
+				tickerLazyLoadPersist := time.NewTicker(time.Minute)
+				defer tickerLazyLoadPersist.Stop()
+
+				lazyLoadC = tickerLazyLoadPersist.C
+			}
+
 			for {
 				select {
 				case <-p.close:
 					return
 				case <-tickerIdleReader.C:
 					p.closeIdleReaders()
-				case <-tickerLazyLoadPersist.C:
+				case <-lazyLoadC:
```

Comment:
Just realised one more thing: do we need to wait until we've finished eagerly loading index headers before starting to persist snapshots of the headers that are loaded?

Reply:
We want to have everything eagerly loaded before starting everything else. Mainly we want to have everything eagerly loaded before service status becomes `ready`.

Reply:
Yes, I agree. If I'm following the code correctly, we start this loop when we create the `ReaderPool`, which starts persisting snapshots, and then later we start eagerly loading index headers. Couldn't we end up in a state where we persist an empty or incomplete snapshot after starting because we haven't finished loading all the eagerly loaded index headers?

Reply:
@charleskorn `func (r *LazyBinaryReader) EagerLoad()` will be called from `func (g *StoreGateway) starting(ctx context.Context) (err error)`. The periodic lazyLoad persist ticker will only be called from `func (g *StoreGateway) running(ctx context.Context) error`. I don't go deep to this service lifecycle codes but I belief, `running` will only be executed after all process in `starting` are completed. I will try to look one more time and see it how it behaves if run  locally.

Reply:
I think now that we immediately read and delete the snapshot file, my original concern is resolved.